File: Pilha/reversestack.c
#include <stdio.h>
#include <stdlib.h>
#define SIZE 10

// Estrutura encadeada
typedef struct No {
    int dado;
    struct No* prox;
} No;

// Estrutura da pilha encadeada
typedef struct Pilha {
    No* topo;
} Pilha;

//Estrutura da fila encadeada
typedef struct Fila {
    No* frente;
    No* fim;
} Fila;

//Pilha
void inicializar(Pilha* pilha) {
    pilha->topo = NULL; // Correção: inicializa o topo com NULL
}

int cheia() {
    return 0; // Como é uma pilha simplesmente encadeada, nunca estará cheia
}

int vazia(Pilha* pilha) {
    return pilha->topo == NULL;
}

void insere(Pilha* pilha, int dado) {
    if(cheia()) { // Correção: chamada da função cheia()
        printf("\nPilha cheia!\n");
        return;
    }

    No* no = (No*)malloc(sizeof(No));
    no->dado = dado;
    no->prox = pilha->topo;
    pilha->topo = no; 
}

int remover(Pilha* pilha) {
    if(vazia(pilha)) {
        printf("\nPilha está vazia!\n");
        return -1;
    }
    No* temp = pilha->topo;
    int dado = temp->dado;
    pilha->topo = temp->prox; // Correção: atualiza o topo corretamente
    free(temp);
    return dado;
}

int topo(Pilha* pilha) {
    if(vazia(pilha)) {
        printf("\nPilha vazia!\n");
        return -1;
    }

    return pilha->topo->dado;
}

//Fila
void inicializafila(Fila* fila) {
    fila->frente = NULL;
    fila->fim = NULL;
}

int filavazia(Fila* fila){
    return fila->frente == NULL;
}

void enfileirar(Fila* fila, int dado) {
    No* no = (No*)malloc(sizeof(No));
    no->dado = dado;
    no->prox = NULL;
    if(fila->fim == NULL) {
        fila->frente = no;
        fila->fim = no;
    } else {
        fila->fim->prox = no;
        fila->fim = no;
    }
}

int desenfileirar(Fila* fila) {
    if(filavazia(fila)) {
        printf("\nFila vazia!\n");
        return -1;
    }
    No* saida = fila->frente;
    int dado = saida->dado;
    fila->frente = saida->prox;
    if(fila->frente == NULL) {
        fila->fim = NULL;
    }
    free(saida);
    return dado;
}

//Invertendo a pilha por meio da fila
void inverter(Pilha* pilha) {
    if(vazia(pilha)){
        printf("\nPilha vazia\n");
        return;
    }

    No* temp = NULL;
    Fila* fila;
    inicializafila(&fila);
    No* atual = pilha->topo;

    //Passo 1: Enfileira todos os dados da pilha na fila
    while(atual != NULL) {
        enfileirar(&fila, atual->dado);
        atual = atual->prox;
    }
    
    //Passo 2: Desenfileira todos os elementos da fila para inverter a ordem
    while(!filavazia(&fila)) {
        int desenfileira = desenfileirar(&fila);
        No* no = (No*)malloc(sizeof(No));
        no->dado = desenfileira;
        no->prox = pilha->topo;
        pilha->topo = no;
    }
}

void imprimirPilha(Pilha* pilha) {
    if(vazia(pilha)) {
        printf("\nPilha vazia!\n");
        return;
    }

    No* atual = pilha->topo;
    printf("\nPilha: ");
    while(atual != NULL) {
        printf("%d ", atual->dado);
        atual = atual->prox;
    }
    printf("\n");
}

void imprimirPilhaInvertida(Pilha* pilha) {
    if(vazia(pilha)) {
        printf("\nPilha vazia!\n");
        return;
    }

    Pilha pilhaAuxiliar;
    inicializar(&pilhaAuxiliar);

    // Copia os elementos da pilha original para uma pilha auxiliar invertendo a ordem
    No* atual = pilha->topo;
    while(atual != NULL) {
        insere(&pilhaAuxiliar, atual->dado);
        atual = atual->prox;
    }

    // Imprime a pilha invertida
    printf("\nPilha Invertida: ");
    atual = pilhaAuxiliar.topo;
    while(atual != NULL) {
        printf("%d ", atual->dado);
        atual = atual->prox;
    }
    printf("\n");
}

int main() {
    Pilha pilha;
    inicializar(&pilha);
    int opcao, dado;

    do {
        printf("\nMenu de Opcoes:\n");
        printf("1 - Inserir\n");
        printf("2 - Remover\n");
        printf("3 - Inverter\n");
        printf("4 - Imprimir Topo\n");
        printf("5 - Imprimir Pilha Toda\n");
        printf("6 - Imprimir Pilha Invertida\n");
        printf("7 - Sair\n");
        printf("Digite a opcao desejada: ");
        scanf("%d", &opcao);

        switch(opcao) {
            case 1:
                printf("Digite o valor a ser inserido: ");
                scanf("%d", &dado);
                insere(&pilha, dado);
                break;
            case 2:
                remover(&pilha);
                break;
            case 3:
                inverter(&pilha);
                printf("Pilha invertida com sucesso!\n");
                break;
            case 4:
                dado = topo(&pilha);
                if(dado != -1)
                    printf("Topo: %d\n", dado);
                break;
            case 5:
                imprimirPilha(&pilha);
                break;
            case 6:
                imprimirPilhaInvertida(&pilha);
                break;
            case 7:
                printf("Encerrando...\n");
                break;
            default:
                printf("Opcao invalida! Tente novamente.\n");
                exit(0);
        }
    } while(opcao != 7);

    return 0;
}

File: Pilha/reversestack.rs
use std::ptr;

// Estrutura encadeada
struct No {
    dado: i32,
    prox: Option<Box<No>>,
}

// Estrutura da pilha encadeada
struct Pilha {
    topo: Option<Box<No>>,
}

// Estrutura da fila encadeada
struct Fila {
    frente: Option<Box<No>>,
    fim: Option<Box<No>>,
}

// Funções da Pilha
fn inicializar(pilha: &mut Pilha) {
    pilha.topo = None;
}

fn cheia() -> bool {
    false
}

fn vazia(pilha: &Pilha) -> bool {
    pilha.topo.is_none()
}

fn insere(pilha: &mut Pilha, dado: i32) {
    if cheia() {
        println!("\nPilha cheia!");
        return;
    }

    let mut no = Box::new(No {
        dado: dado,
        prox: pilha.topo.take(),
    });

    pilha.topo = Some(no);
}

fn remover(pilha: &mut Pilha) -> i32 {
    if vazia(pilha) {
        println!("\nPilha está vazia!");
        return -1;
    }

    let temp = pilha.topo.take().unwrap();
    let dado = temp.dado;
    pilha.topo = temp.prox;
    dado
}

fn topo(pilha: &Pilha) -> i32 {
    if vazia(pilha) {
        println!("\nPilha vazia!");
        return -1;
    }

    pilha.topo.as_ref().unwrap().dado
}

// Funções da Fila
fn inicializafila(fila: &mut Fila) {
    fila.frente = None;
    fila.fim = None;
}

fn filavazia(fila: &Fila) -> bool {
    fila.frente.is_none()
}

fn enfileirar(fila: &mut Fila, dado: i32) {
    let no = Box::new(No {
        dado: dado,
        prox: None,
    });

    if fila.fim.is_none() {
        fila.frente = Some(no.clone());
        fila.fim = Some(no);
    } else {
        fila.fim.as_mut().unwrap().prox = Some(no);
        fila.fim = Some(no);
    }
}

fn desenfileirar(fila: &mut Fila) -> i32 {
    if filavazia(fila) {
        println!("\nFila vazia!");
        return -1;
    }

    let saida = fila.frente.take().unwrap();
    let dado = saida.dado;
    fila.frente = saida.prox;
    if fila.frente.is_none() {
        fila.fim = None;
    }
    dado
}

// Função para inverter a pilha por meio da fila
fn inverter(pilha: &mut Pilha) {
    if vazia(pilha) {
        println!("\nPilha vazia");
        return;
    }

    let mut temp: Option<Box<No>> = None;
    let mut fila = Fila {
        frente: None,
        fim: None,
    };
    let mut atual = pilha.topo.take();

    // Passo 1: Enfileira todos os dados da pilha na fila
    while let Some(mut node) = atual.take() {
        enfileirar(&mut fila, node.dado);
        atual = node.prox.take();
    }
    
    // Passo 2: Desenfileira todos os dados da fila e empilha na pilha temporária
    while let Some(dado) = desenfileirar(&mut fila) {
        insere(&mut temp, dado);
    }

    // Passo 3: Desempilha todos os dados da pilha temporária e empilha na pilha original
    while let Some(dado) = remover(&mut temp) {
        insere(pilha, dado);
    }
}
    // Função principal
fn main() {
    let mut pilha = Pilha {topo: None,};
    // Inicializa a pilha com alguns valores
    insere(&mut pilha, 1);
    insere(&mut pilha, 2);
    insere(&mut pilha, 3);
    insere(&mut pilha, 4);

    println!("Pilha original:");
    while !vazia(&pilha) {
       println!("{}", remover(&mut pilha));
}

    inverter(&mut pilha);

    println!("Pilha invertida:");
    while !vazia(&pilha) {
       println!("{}", remover(&mut pilha));
    }
}
